google/cloud/logging_v2/_instrumentation.py

```diff
@@ -0,0 +1,173 @@
+    return new_entries
+
+
+def _create_diagnostic_entry(name=_PYTHON_LIBRARY_NAME, version=_LIBRARY_VERSION, **kw):
+    """Create a diagnostic log entry describing this library
+
+        The diagnostic log consists of a list of library name and version objects
+        that have handled a given log entry.  If this library is the originator
+        of the log entry, it will look like:
+        {logging.googleapis.com/diagnostic: {instrumentation_source: [{name: "python", version: "3.0.0"}]}}
+
+    Args:
+        name(str): The name of this library (e.g. 'python')
+        version(str) The version of this library (e.g. '3.0.0')
+
+    Returns:
+        google.cloud.logging_v2.LogEntry: Log entry with library information
+    """
+    payload = {
+        _DIAGNOSTIC_INFO_KEY: {
+            _INSTRUMENTATION_SOURCE_KEY: [_get_instrumentation_source(name, version)]
+        }
+    }
+    kw["severity"] = "INFO"
+    entry = StructEntry(payload=payload, **kw)
+    return entry
+
+
+def _truncate_string(str, max_length):
+    """Truncate a string to a maximum length
+
+    Args:
+        str(str): The string to truncate
+        max_length(int): The maximum length
+
+    Returns:
+        A string containing either 'str' or a truncated version of
+        'str' with an asterisk at the end
+    """
+    if len(str) > max_length:
+        return str[:max_length] + "*"
+    else:
+        return str
+
+
+def _validate_and_update_instrumentation(existing_info=None):
+    """Validate existing instrumentation info and append the final entry
+
+        Validates that existing instrumentation info matches this
+        library's info, and ensures that there are at most 3 entries
+        in the instrumentation_source array (to avoid malicious log inflation).
+
+    Args:
+        existing_info: A list of instrumentation_source objects already on an etnry
+
+    Returns:
+        A validated list of instrumentation_source objects with the current
+        library entry at the end
+    """
+    info_stack = []
+    info_stack.append(_get_instrumentation_source())
+    if existing_info:
+        for info in existing_info[::-1]:
+            if len(info_stack) == 3:
+                break
+            if _is_valid(info):
+                info_stack.append(
+                    _get_instrumentation_source(info["name"], info["version"])
+                )
+
+    info_stack.reverse()
+    return info_stack
+
+
+def _get_instrumentation_source(name=_PYTHON_LIBRARY_NAME, version=_LIBRARY_VERSION):
+    """Gets a JSON representation of the instrumentation_source
+
+    Args:
+        name(str): The name of this library (e.g. 'python')
+        version(str) The version of this library (e.g. '3.0.0')
+    Returns:
+       obj: JSON object with library information
+    """
+    return {
+        "name": _truncate_string(name, _MAX_NAME_LENGTH),
+        "version": _truncate_string(version, _MAX_VERSION_LENGTH),
```

Comment:
suggestion: since the truncation logic can be written as a simple one-liner, and since it's used in one place, and since there's nothing else going on in this function, I'd probably suggest just doing the truncation in-place here:

```
def _get_instrumentation_source(name=_PYTHON_LIBRARY_NAME, version=_LIBRARY_VERSION):
    source = {"name": name, "version": version}
   # truncate strings to no more than _MAX_NAME_LENGTH characters
   for key, val in source.items():
     source[key] = val if len(val) <= _MAX_NAME_LENGTH else f"{val[:_MAX_NAME_LENGTH]}*"
```

But if you think it'll be useful in other places later, it's probably better to keep it as a separate function

Reply:
That looks great.  Thanks for the suggestion.